### ISSUE TITLE
NimBLE init and deinit memory leak  (IDFGH-6480)

### ISSUE THREAD

**Issue by kadosch** [espressif/esp-idf#8136]
## Environment

- Development Kit:     Olimex ESP32-DevKit-LiPo
- Kit version: rev.C
- Module or chip used:  ESP32-WROOM-32E
- IDF version: v4.3.1-480-g233dc30fb1
- Build System:         idf.py
- Compiler version: xtensa-esp32-elf-gcc (crosstool-NG esp-2021r2) 8.4.0
- Operating System:     Linux
- Using an IDE?: No
- Power Supply:         USB

## Problem Description

I'm working on a project that uses NimBLE and enters light sleep mode periodically, so it has to disable the bluetooth controller before going to sleep and reenable it after waking up.

There seems to be a memory leak in the init/deinit code from the ESP-IDF NimBLE port, as my device eventually runs out of heap memory.

The problem can be reproduced with the provided minimal example.

### Expected Behavior
NimBLE bluetooth controller is initialized and deinitlialized without memory leaks.

### Actual Behavior
NimBLE bluetooth controller is initialized and deinitlialized with memory leaks.

### Steps to reproduce
Simply execute the provided code.

### Code to reproduce this issue

```cpp
#include <esp_log.h>
#include <nvs_flash.h>
#include <esp_bt.h>
#include <esp_nimble_hci.h>
#include <nimble/nimble_port.h>
#include <nimble/nimble_port_freertos.h>
#include <esp_heap_trace.h>

#define NUM_RECORDS 100
static heap_trace_record_t trace_record[NUM_RECORDS];
static const char *TAG = "nimble_leak_test";

static void ble_manager_host_task(void *param)
{
    (void)param;

    ESP_LOGI(TAG, "BLE Host Task Started");
    /* This function will return only when nimble_port_stop() is executed */
    nimble_port_run();

    nimble_port_freertos_deinit();
}

void start_ble_func()
{
    esp_err_t err = esp_nimble_hci_and_controller_init();
    if (err != ESP_OK)
    {
        ESP_LOGE(TAG, "Error at nimble controller init: %s", esp_err_to_name(err));
        return;
    }

    nimble_port_init();

    nimble_port_freertos_init(ble_manager_host_task);
}

void stop_ble_func()
{
    int rc = nimble_port_stop();
    if (rc != 0)
    {
        ESP_LOGE(TAG, "Error at nimble port stop");
        return;
    }

    nimble_port_deinit();

    esp_err_t err = esp_nimble_hci_and_controller_deinit();
    if (err != ESP_OK)
    {
        ESP_LOGE(TAG, "Error at nimble controller deinit: %s", esp_err_to_name(err));
        return;
    }
}

void app_main(void)
{
    esp_err_t ret;

    // Initialize NVS.
    ret = nvs_flash_init();
    if (ret == ESP_ERR_NVS_NO_FREE_PAGES || ret == ESP_ERR_NVS_NEW_VERSION_FOUND)
    {
        ESP_ERROR_CHECK(nvs_flash_erase());
        ret = nvs_flash_init();
    }
    ESP_ERROR_CHECK(ret);

    ESP_ERROR_CHECK(heap_trace_init_standalone(trace_record, NUM_RECORDS));

    int free_heap_size = esp_get_free_heap_size();

    for (int i = 0; i < 5; i++)
    {
        free_heap_size = esp_get_free_heap_size();
        ESP_LOGI(TAG, "free_heap_size %d ", free_heap_size);

        start_ble_func();
        free_heap_size = esp_get_free_heap_size();
        ESP_LOGI(TAG, "after start_ble_func %d ", free_heap_size);
        vTaskDelay(pdMS_TO_TICKS(500));
        stop_ble_func();
        free_heap_size = esp_get_free_heap_size();
        ESP_LOGI(TAG, "after stop_ble_func %d \n", free_heap_size);
    }

    ESP_ERROR_CHECK(heap_trace_start(HEAP_TRACE_LEAKS));

    start_ble_func();

    vTaskDelay(pdMS_TO_TICKS(1000));

    stop_ble_func();

    ESP_ERROR_CHECK(heap_trace_stop());
    heap_trace_dump();

    return;
}


```

## Debug Logs

```
I (513) nimble_leak_test: free_heap_size 224892 
I (513) BTDM_INIT: BT controller compile version [6a07b06]
I (513) system_api: Base MAC address is not set
I (523) system_api: read default base MAC address from EFUSE
I (533) phy_init: phy_version 4670,719f9f6,Feb 18 2021,17:07:07
I (863) nimble_leak_test: BLE Host Task Started
I (863) nimble_leak_test: after start_ble_func 205796 
I (1363) NimBLE: GAP procedure initiated: stop advertising.

W (1363) Timer: Timer not stopped
I (1363) nimble_leak_test: after stop_ble_func 234952 

I (1363) nimble_leak_test: free_heap_size 234952 
I (1373) BTDM_INIT: BT controller compile version [6a07b06]
I (1613) nimble_leak_test: BLE Host Task Started
I (1613) nimble_leak_test: after start_ble_func 202388 
I (2113) NimBLE: GAP procedure initiated: stop advertising.

W (2113) Timer: Timer not stopped
I (2113) nimble_leak_test: after stop_ble_func 234884 

I (2113) nimble_leak_test: free_heap_size 234884 
I (2123) BTDM_INIT: BT controller compile version [6a07b06]
I (2363) nimble_leak_test: BLE Host Task Started
I (2373) nimble_leak_test: after start_ble_func 202324 
I (2873) NimBLE: GAP procedure initiated: stop advertising.

W (2873) Timer: Timer not stopped
I (2873) nimble_leak_test: after stop_ble_func 234808 

I (2873) nimble_leak_test: free_heap_size 234808 
I (2883) BTDM_INIT: BT controller compile version [6a07b06]
I (3133) nimble_leak_test: BLE Host Task Started
I (3133) nimble_leak_test: after start_ble_func 202260 
I (3633) NimBLE: GAP procedure initiated: stop advertising.

W (3633) Timer: Timer not stopped
I (3633) nimble_leak_test: after stop_ble_func 234740 

I (3633) nimble_leak_test: free_heap_size 234740 
I (3643) BTDM_INIT: BT controller compile version [6a07b06]
I (3883) nimble_leak_test: BLE Host Task Started
I (3883) nimble_leak_test: after start_ble_func 202192 
I (4383) NimBLE: GAP procedure initiated: stop advertising.

W (4383) Timer: Timer not stopped
I (4383) nimble_leak_test: after stop_ble_func 234676 

I (4383) BTDM_INIT: BT controller compile version [6a07b06]
I (4633) nimble_leak_test: BLE Host Task Started
I (5633) NimBLE: GAP procedure initiated: stop advertising.

W (5633) Timer: Timer not stopped
4 allocations trace (100 entry buffer)
32 bytes (@ 0x3ffcb390) allocated CPU 0 ccount 0x2be69f60 caller 0x400d5640:0x400de164:0x400d6ed5:0x400d73a7:0x400d5f81:0x400d5dc4:0x400d5f02:0x4011efa8:0x40093b50:
0x400d5640: esp_timer_create at /home/javi/esp/esp-idf/components/esp_timer/src/esp_timer.c:114

0x400de164: npl_freertos_callout_init at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/porting/npl/freertos/src/npl_os_freertos.c:361

0x400d6ed5: ble_hs_stop_init at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/nimble/host/src/ble_hs_stop.c:283

0x400d73a7: ble_hs_init at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/nimble/host/src/ble_hs.c:775 (discriminator 2)

0x400d5f81: nimble_port_init at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/porting/nimble/src/nimble_port.c:53

0x400d5dc4: start_ble_func at /tmp/nimble_leak_test/main/nimble_leak_test.c:35

0x400d5f02: app_main at /tmp/nimble_leak_test/main/nimble_leak_test.c:92 (discriminator 2)

0x4011efa8: main_task at /home/javi/esp/esp-idf/components/freertos/port/port_common.c:134 (discriminator 2)

0x40093b50: vPortTaskWrapper at /home/javi/esp/esp-idf/components/freertos/port/xtensa/port.c:170

4096 bytes (@ 0x3ffc9590) allocated CPU 0 ccount 0x2be6f810 caller 0x4009220b:0x400d5f3e:0x400d5dca:0x400d5f02:0x4011efa8:0x40093b50:
0x4009220b: xTaskCreatePinnedToCore at /home/javi/esp/esp-idf/components/freertos/tasks.c:830

0x400d5f3e: nimble_port_freertos_init at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/porting/npl/freertos/src/nimble_port_freertos.c:51

0x400d5dca: start_ble_func at /tmp/nimble_leak_test/main/nimble_leak_test.c:36

0x400d5f02: app_main at /tmp/nimble_leak_test/main/nimble_leak_test.c:92 (discriminator 2)

0x4011efa8: main_task at /home/javi/esp/esp-idf/components/freertos/port/port_common.c:134 (discriminator 2)

0x40093b50: vPortTaskWrapper at /home/javi/esp/esp-idf/components/freertos/port/xtensa/port.c:170

344 bytes (@ 0x3ffcb4c8) allocated CPU 0 ccount 0x2be70618 caller 0x40092219:0x400d5f3e:0x400d5dca:0x400d5f02:0x4011efa8:0x40093b50:
0x40092219: xTaskCreatePinnedToCore at /home/javi/esp/esp-idf/components/freertos/tasks.c:835

0x400d5f3e: nimble_port_freertos_init at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/porting/npl/freertos/src/nimble_port_freertos.c:51

0x400d5dca: start_ble_func at /tmp/nimble_leak_test/main/nimble_leak_test.c:36

0x400d5f02: app_main at /tmp/nimble_leak_test/main/nimble_leak_test.c:92 (discriminator 2)

0x4011efa8: main_task at /home/javi/esp/esp-idf/components/freertos/port/port_common.c:134 (discriminator 2)

0x40093b50: vPortTaskWrapper at /home/javi/esp/esp-idf/components/freertos/port/xtensa/port.c:170

32 bytes (@ 0x3ffcb648) allocated CPU 0 ccount 0x2c2ed1b0 caller 0x400d5640:0x400de164:0x400dba5e:0x400d9c8e:0x400d70c7:0x400d7286:0x400d728e:0x400d5fb1:0x400d5d8d:0x40093b50
0x400d5640: esp_timer_create at /home/javi/esp/esp-idf/components/esp_timer/src/esp_timer.c:114

0x400de164: npl_freertos_callout_init at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/porting/npl/freertos/src/npl_os_freertos.c:361

0x400dba5e: ble_hs_flow_startup at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/nimble/host/src/ble_hs_flow.c:275

0x400d9c8e: ble_hs_startup_go at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/nimble/host/src/ble_hs_startup.c:367

0x400d70c7: ble_hs_sync at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/nimble/host/src/ble_hs.c:345

0x400d7286: ble_hs_start at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/nimble/host/src/ble_hs.c:668

0x400d728e: ble_hs_event_start_stage2 at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/nimble/host/src/ble_hs.c:569

0x400d5fb1: nimble_port_run at /home/javi/esp/esp-idf/components/bt/host/nimble/nimble/porting/nimble/src/nimble_port.c:79

0x400d5d8d: ble_manager_host_task at /tmp/nimble_leak_test/main/nimble_leak_test.c:21

0x40093b50: vPortTaskWrapper at /home/javi/esp/esp-idf/components/freertos/port/xtensa/port.c:170

4504 bytes 'leaked' in trace (4 allocations)
total allocations 38 total frees 41
```

## Other items if possible
[nimble_leak_test.zip](https://github.com/espressif/esp-idf/files/7762880/nimble_leak_test.zip)

**Comment by rahult-github** [comment]
Hi @kadosch ,
  Very recently, during fixing one of the issue https://github.com/espressif/esp-idf/issues/7653 , we have fixed memory leak during init / deinit.  Can you please confirm the commit https://github.com/espressif/esp-idf/commit/0f96ebce139ff9fc63069819e0e568adac8a6e3f exists at your end ? 

Thanks,
 Rahul

**Comment by kadosch** [comment]
Hi @rahult-github,
yeah, I saw that issue before opening this one, but they are different. I'm not using power management at all, and the commit that fixes that issue is already included in the branch that I'm working with:

```console
[javi@LEGION esp-idf]$ git branch --show-current
release/v4.3
[javi@LEGION esp-idf]$ git describe --tags
v4.3.1-480-g233dc30fb1
[javi@LEGION esp-idf]$ git branch --contains 0f96ebc
* release/v4.3
```

Regards.

**Comment by kadosch** [comment]
Hi again,
recent commit 7a4ed65769bd4c4dfafdb4c85d655bab70a13c18 fixes the problem. Funnily enough, that commit was merged to branch release/v4.3 barely 20 minutes after creating this issue.

Anyway, closing the issue as resolved.
